File: go/tests/hr_board_test.go
// hr_board_test.go
//
// Verifies the CircleAI.HR port (hr_board.go): hire/get, Employees name-ordering,
// leave request + decide (unknown-id error) + pending filter, and AvgRatingFor
// (mean over reviews; 0.0 when none).

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestHR_HireAndEmployeesOrdered(t *testing.T) {
	b := circleai.NewInMemoryHRBoard()
	b.Hire(circleai.Employee{EmployeeId: "e1", Name: "Zola", Role: "Dev", Salary: circleai.DecimalFromInt(50000), Currency: "ZAR"})
	b.Hire(circleai.Employee{EmployeeId: "e2", Name: "Ada", Role: "PM", Salary: circleai.DecimalFromInt(60000), Currency: "ZAR"})

	if e, ok := b.GetEmployee("e1"); !ok || e.Name != "Zola" {
		t.Fatalf("get e1 = %+v ok=%v", e, ok)
	}
	if _, ok := b.GetEmployee("ghost"); ok {
		t.Fatalf("unknown employee must be absent")
	}
	emps := b.Employees()
	if len(emps) != 2 || emps[0].Name != "Ada" || emps[1].Name != "Zola" {
		t.Fatalf("employees name order wrong: %+v", emps)
	}
}

func TestHR_LeaveDecisionAndPending(t *testing.T) {
	b := circleai.NewInMemoryHRBoard()
	from := time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC)
	b.Request(circleai.LeaveRequest{RequestId: "r1", EmployeeId: "e1", Kind: "Annual", From: from, To: from.AddDate(0, 0, 5), Status: "Pending"})
	b.Request(circleai.LeaveRequest{RequestId: "r2", EmployeeId: "e2", Kind: "Sick", From: from, To: from, Status: "pending"})

	if len(b.PendingLeaves()) != 2 {
		t.Fatalf("pending should be 2")
	}
	if err := b.DecideLeave("r1", "Approved"); err != nil {
		t.Fatalf("decide: %v", err)
	}
	if err := b.DecideLeave("ghost", "Approved"); err == nil {
		t.Fatalf("unknown leave request must error")
	}
	pend := b.PendingLeaves()
	if len(pend) != 1 || pend[0].RequestId != "r2" {
		t.Fatalf("pending after decision wrong: %+v", pend)
	}
}

func TestHR_AvgRating(t *testing.T) {
	b := circleai.NewInMemoryHRBoard()
	when := time.Date(2026, 6, 1, 0, 0, 0, 0, time.UTC)
	b.Review(circleai.PerformanceReview{ReviewId: "v1", EmployeeId: "e1", ReviewedOn: when, RatingOutOf5: 4})
	b.Review(circleai.PerformanceReview{ReviewId: "v2", EmployeeId: "e1", ReviewedOn: when, RatingOutOf5: 2})
	if got := b.AvgRatingFor("e1"); got != 3.0 {
		t.Fatalf("avg rating = %v, want 3.0", got)
	}
	// No reviews -> 0.0 (DefaultIfEmpty(0).Average()).
	if got := b.AvgRatingFor("nobody"); got != 0.0 {
		t.Fatalf("avg rating no-reviews = %v, want 0.0", got)
	}
}

File: typescript/tests/hr_board.test.ts
// hr_board.test.ts
// Verifies the CircleAI.HR port: employee registry (name-ordered), leave
// requests + decisions, performance reviews with average rating, and context.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryHRBoard,
  HRDomainContext,
  employee,
  leaveRequest,
  performanceReview,
} from "../src/hr/index";

const D = (s: string) => new Date(s);

describe("InMemoryHRBoard", () => {
  it("hires and lists employees ordered by name; unknown id undefined", () => {
    const b = new InMemoryHRBoard();
    b.hire(employee("e2", "Zoe", "Dev", D("2020-01-01"), 100, "ZAR"));
    b.hire(employee("e1", "Ada", "Lead", D("2019-01-01"), 200, "ZAR"));
    assert.deepEqual(
      b.employees.map((e) => e.name),
      ["Ada", "Zoe"],
    );
    assert.equal(b.getEmployee("e1")?.role, "Lead");
    assert.equal(b.getEmployee("nope"), undefined);
  });

  it("tracks pending leave and applies decisions; unknown throws", () => {
    const b = new InMemoryHRBoard();
    b.request(leaveRequest("l1", "e1", "Annual", D("2026-01-01"), D("2026-01-05"), "Pending"));
    b.request(leaveRequest("l2", "e2", "Sick", D("2026-02-01"), D("2026-02-02"), "Approved"));
    assert.deepEqual(
      b.pendingLeaves().map((l) => l.requestId),
      ["l1"],
    );
    b.decideLeave("l1", "Approved");
    assert.deepEqual(b.pendingLeaves(), []);
    assert.throws(() => b.decideLeave("ghost", "X"), /Unknown leave request ghost/);
  });

  it("averages ratings; empty is 0", () => {
    const b = new InMemoryHRBoard();
    assert.equal(b.avgRatingFor("e1"), 0);
    b.review(performanceReview("r1", "e1", D("2026-01-01"), 4, ""));
    b.review(performanceReview("r2", "e1", D("2026-06-01"), 2, ""));
    b.review(performanceReview("r3", "e2", D("2026-01-01"), 5, ""));
    assert.equal(b.avgRatingFor("e1"), 3);
    assert.equal(b.avgRatingFor("e2"), 5);
  });

  it("rejects null arguments", () => {
    const b = new InMemoryHRBoard();
    assert.throws(() => b.hire(null as never));
    assert.throws(() => b.request(null as never));
    assert.throws(() => b.review(null as never));
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(HRDomainContext.systemPromptSnippet.includes("[DOMAIN: HR]"));
    assert.deepEqual(HRDomainContext.complianceFlags, ["LRA_66_1995", "BCEA", "EEA", "Skills_Development_Act", "POPIA"]);
    assert.deepEqual(HRDomainContext.suggestedTools, ["hris", "document_editor", "analytics", "job_boards"]);
  });
});
